lib/improver/argparser.py

```diff
@@ -28,19 +28,185 @@
+              Default is None, which does not add additional arguments.
+            kwargs (dictionary):
+              Additional keyword arguments which are passed to the superclass
+              constructor (argparse.ArgumentParser).
+        """
+
+        self._args = None
+
+        # Allow either central_arguments or specific_arguments to be None
+        # (or empty lists)
+        if central_arguments is None:
+            central_arguments = []
+        if specific_arguments is None:
+            specific_arguments = []
+
+        # argspecs of the compulsory arguments (no switch here)
+        compulsory_arguments = ArgParser.COMPULSORY_ARGUMENTS.values()
+
+        # get argspecs of the central arguments
+        central_arguments = self._args_from_selected_central_args(
+                             central_arguments)
+
+        # specific arguments must be passed in with the correct format
+        # (argspecs)- we don't need to do anything special here...
+
+        # create instance of ArgumentParser (pass along kwargs)
+        super(ArgParser, self).__init__(**kwargs)
+
+        cli_arguments = (compulsory_arguments + central_arguments +
+                         specific_arguments)
+
+        # automatically add all of the arguments
+        self.add_arguments(cli_arguments)
+        # Done. Now we can get the arguments with self.args()
+
+    @staticmethod
+    def _args_from_selected_central_args(arg_name_list):
+        """Return a list of argument specifications from the
+        ArgParser.CENTRALIZED_ARGUMENTS, given a list of keys (argument names).
+
+        Args:
+            arg_name_list (list):
+              A list containing the keys required from the centralized argument
+              dictionary.
+
+        Returns:
+            list:
+              The argument specifications which may later be used to add
+              arguments to the parser.
+        """
+        return [ArgParser.CENTRALIZED_ARGUMENTS[arg_name] for arg_name
+                in arg_name_list]
+
+    def add_arguments(self, argspec_list):
+        """Adds a list of arguments to the ArgumentParser.
+
+        The input argspec_list is a list of argument specifications, where each
+        element (argument specification) is a tuple/list of length 1 or 2.
+        The first element of an argument specification is a list of strings
+        which the name/flags used to add the argument.
+        (Optionally) the second element of the argument spec shall be a
+        dictionary containing the keyword arguments which are passed into the
+        add_arguments() method.
+
+        Args:
+            argspec_list (list):
+              A list containing the specifications required to add the
+              arguments (see above)
+
+        Raises:
+            AttributeError:
+                Notifies the user if any of the argument specifications has
+                the wrong length (not 1 or 2).
+        """
```

Comment:
We had some discussion about this.  My understanding is that there are two benefits of this function over using the in-built "add_argument" function from argparse: one is the ability to accept an empty keyword-args dict, and one is in raising the attribute error.  I would argue that we should not permit the first, as every argument to IMPROVER CLIs should have at least an associated "help" string, so the dict should never be empty.  I also find it quite unclear why the attribute error is raised here - it's not obvious from the code what conditions would lead to the error.  So I think this function adds too much complexity than the benefits it provides.

For simplicity and maintainability, I'd like to see this moved back out into the "init" function, using the existing argparse "add_argument" function to add each of the specified arguments.

Reply:
I think this is mainly for checking the definitions that come from the specific_arguments from the CLIs. This could be got rid of if we go back to the old way of adding arguments in the CLI, as the defaults in the ArgParser itself should be set up correctly. So I agree with Caroline, we could probably get ride of this.

Reply:
I'm happy to remove the option of allowing for an argspec with length 1 (effectively requiring that the kwargs are always supplied, even if we choose to pass in an empty dict (which is unlikely, because we always require a `help`) - but I do think we should check for an unexpected length argspec, because it's good defensive programming (e.g: what if someone mistakenly defined one of the centralized arguments in a bad way? - we could rely on a different python runtime error, but I think the error message might be hard to decipher)

As for whether the method is required at all - You're right, if we decided only to add a single argument at a time, we could get rid of this entirely. That said, I do think that it's good practice to break up a method into several smaller methods - as this makes it easier to test, and encourages code reuse.

Reply:
I do see your point about breaking up "init".  For that purpose then, could we add a leading underscore to the function name please, to indicate that this should only be called internally for initialising the class.

Reply:
I think if we're allowing for `add_argument()` to be called (by virtue of inheriting from `ArgumentParser`), it also makes sense for `add_arguments()` to be called.

Reply:
Okay, fair enough. I've removed the treatment of length 1 tuples/lists.

(Enforcing a length of 2 does mean that people can just put in an empty dictionary, but I think this is what we want - we could enforce other things on the dictionary (e.g: that it always has a help), but I think that would create a further divergence in functionality between this and `add_argument`.)